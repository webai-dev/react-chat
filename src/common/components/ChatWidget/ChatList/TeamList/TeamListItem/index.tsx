import React, { useContext, useState } from "react";
import { makeStyles } from "@material-ui/styles";
import Accordion from "@material-ui/core/Accordion";
import AccordionSummary from "@material-ui/core/AccordionSummary";
import AccordionDetails from "@material-ui/core/AccordionDetails";
import Box from "@material-ui/core/Box";
import Checkbox from "@material-ui/core/Checkbox";
import Typography from "@material-ui/core/Typography";
import ExpandMoreIcon from "@material-ui/icons/ExpandMore";
import PeopleIcon from "@material-ui/icons/People";
import NullableImage from "../../../_Common/NullableImage";
import { TeamItem, TeamMember } from "../../../../../../types/types";
import { CompanyContext } from "../../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../../hooks/useChatRooms";
import Chat from "../../../../../services/chat";
// import { createTeamChat } from "../../../../../services/teams";

// Global styles
export const useStyles = makeStyles({
  accordion: {
    margin: "0 0 -2px 0 !important",
    borderTop: "1px solid rgba(0, 0, 0, 0.1)",
    borderBottom: "1px solid rgba(0, 0, 0, 0.1)",
    borderRadius: "0 !important",
  },
  accordionDetails: {
    padding: 0,
  },
  accordionSummary: {
    padding: "16px 32px",
    borderBottom: "1px solid rgba(0, 0, 0, 0.1)",
  },
  iconContainer: {
    paddingLeft: "10px",
    paddingRight: "5px",
    color: "#707070",
  },
  info: {
    marginLeft: 50,
  },
  name: {
    color: "#707070",
    fontSize: 16,
  },
  occupation: {
    color: "#707070",
    fontSize: 13,
  },
  photo: {
    width: 50,
    height: 50,
    borderRadius: 25,
  },
  root: {
    position: "relative",
    display: "flex",
    alignItems: "center",
    padding: "16px 22px 13px",
    cursor: "pointer",
    backgroundColor: (selected) => (selected ? "#4F8DCB1A" : "white"),
  },
  status: {
    position: "absolute",
    width: 12,
    height: 12,
    bodorRadius: 8,
    boxShadow: "0px 1px 3px #00000029",
    border: "1px solid #707070",
    borderRadius: 6,
    bottom: 0,
    right: 0,
  },
  unread: {
    width: 3,
    height: "100%",
    position: "absolute",
    backgroundColor: "#3870A0",
    left: 0,
    top: 0,
  },
});

interface TeamListItemProps {
  modalItem?: boolean;
  team: TeamItem;
  selectedMembers?: number[];
  setSelectedMembers?: React.Dispatch<React.SetStateAction<number[]>>;
  connectionStatus: number;
}

/**
 * TeamListItem component
 * @param {TeamListItemProps} props Properties
 * @description TeamListItem component consists of data for one user/group in Chat Room
 */
const TeamListItem: React.FC<TeamListItemProps> = ({
  modalItem = false,
  team = {
    id: 0,
    name: "",
    members: [],
    chatRoomID: 0,
  },
  selectedMembers = [],
  setSelectedMembers = () => undefined,
  connectionStatus = 1,
}) => {
  const { name, members } = team;
  const [selectedMemberId, setSelectedMemberId] = useState(-1);
  const classes = useStyles();
  const { addChatRoom } = useContext(ChatRoomsContext);
  const { selectedCompany } = useContext(CompanyContext);
  const { checkDirectChat } = Chat;

  const handleClickMember = async (id: number, userID: number) => {
    if (!modalItem) {
      setSelectedMemberId(id);

      const response = await checkDirectChat(
        selectedCompany && selectedCompany.id,
        userID
      );
      if (response) {
        if (Object.keys(response).length === 0) {
        } else {
          addChatRoom(response.id, selectedCompany && selectedCompany.id);
        }
      }
    }
  };
  const handleClickTeamChat = async (id: number, chatRoomID: number | null) => {
    if (chatRoomID) {
      addChatRoom(chatRoomID, selectedCompany && selectedCompany.id);
    } else {
      // const response = await createTeamChat(
      //   selectedCompany && selectedCompany.id,
      //   id
      // );

      // if (response) {
      //   if (Object.keys(response).length === 0) {
      //     console.log("Empty response");
      //   } else {
      //     console.log(response);
      //     addChatRoom(response.id, selectedCompany && selectedCompany.id);
      //   }
      // }
    }
  };

  return (
    <Accordion data-testid="team-list-item" className={classes.accordion}>
      <AccordionSummary
        className={classes.accordionSummary}
        expandIcon={<ExpandMoreIcon />}
      >
        <Typography>
          <span>
            <span>{name}</span>
          </span>
        </Typography>
      </AccordionSummary>
      <AccordionDetails className={classes.accordionDetails}>
        <Box width="100%">
          {!modalItem && team.isTeamMember && (
            <div
              data-testid="room-list-item"
              className={classes.root}
              style={{
                backgroundColor: selectedMemberId === 0 ? "#4F8DCB1A" : "white",
              }}
              onClick={() => handleClickTeamChat(team.id, team.chatRoomID)}
            >
              <div className={classes.iconContainer}>
                <PeopleIcon fontSize="large" />
              </div>
              <div className={classes.info}>
                <div className={classes.name}>Chat with the team</div>
                <div className={classes.occupation}></div>
              </div>
            </div>
          )}

          {members &&
            members.map((member: TeamMember, index: number) => (
              <div
                data-testid="room-list-item"
                className={classes.root}
                style={{
                  backgroundColor:
                    selectedMemberId === member.id ? "#4F8DCB1A" : "white",
                }}
                onClick={() =>
                  handleClickMember(member.id, member.memberUserID)
                }
                key={index}
              >
                {modalItem && (
                  <Checkbox
                    color="default"
                    checked={
                      selectedMembers.filter((id) => id === member.memberUserID)
                        .length > 0
                    }
                    onChange={(e) => {
                      if (e.target.checked) {
                        setSelectedMembers([
                          ...selectedMembers,
                          member.memberUserID,
                        ]);
                      } else {
                        let updatedMembers = selectedMembers.filter(
                          (id) => id !== member.memberUserID
                        );
                        setSelectedMembers(updatedMembers);
                      }
                    }}
                    inputProps={{ "aria-label": "checkbox with default color" }}
                  />
                )}

                <NullableImage
                  src={
                    member.memberUser ? member.memberUser.profilePhotoUrl : ""
                  }
                  className={classes.photo}
                />
                <div className={classes.info}>
                  <div className={classes.name}>
                    {member.memberUser ? member.memberUser.firstName : ""}{" "}
                    {member.memberUser ? member.memberUser.lastName : ""}
                  </div>
                  <div className={classes.occupation}>
                    {(member.memberUser
                      ? member.memberUser.companyMembership
                      : []
                    )
                      .map((memberShip) => memberShip.jobTitle)
                      .join(", ")}
                  </div>
                </div>
              </div>
            ))}
        </Box>
      </AccordionDetails>
    </Accordion>
  );
};

export default TeamListItem;
