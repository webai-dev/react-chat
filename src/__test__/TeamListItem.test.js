import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import TeamListItem from "../common/components/ChatWidget/ChatList/TeamList/TeamListItem";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const { getByTestId } = render(<TeamListItem />);
  const teamListItem = getByTestId("team-list-item");
  expect(teamListItem).toBeTruthy();
});

it("matches snapshot", () => {
  let modal = false,
    team = {
      id: 0,
      name: "",
      members: [],
    };
  const tree = renderer.create(<TeamListItem team={team} />).toJSON();
  expect(tree).toMatchSnapshot();
});
