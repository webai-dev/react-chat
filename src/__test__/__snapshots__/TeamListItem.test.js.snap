// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="MuiPaper-root MuiAccordion-root makeStyles-accordion-14 MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  data-testid="team-list-item"
>
  <div
    aria-disabled={false}
    aria-expanded={false}
    className="MuiButtonBase-root MuiAccordionSummary-root makeStyles-accordionSummary-16"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiAccordionSummary-content"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        <span>
          <span>
            
          </span>
        </span>
      </p>
    </div>
    <div
      aria-disabled={false}
      aria-hidden={true}
      className="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      onBlur={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      role={null}
      tabIndex={null}
    >
      <span
        className="MuiIconButton-label"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="MuiCollapse-container MuiCollapse-hidden"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      >
        <div
          role="region"
        >
          <div
            className="MuiAccordionDetails-root makeStyles-accordionDetails-15"
          >
            <div
              className="MuiBox-root MuiBox-root-26"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
